concepts/code-organization/introduction.md

```diff
@@ -0,0 +1,161 @@
+
+// Definition of myFunction
+int myFunction(int n) {
+    if (n < 0) {
+        return 0;
+    } else {
+        return myOtherFunction(n - 2);
+    }
+}
+
+// Definition of myOtherFunction
+int myOtherFunction(int m) {
+    return myFunction(m / 2);
+}
+```
+
+Attention: the `;` is needed after the declaration, but not necessarily after the definition.
+
+## Header Files
+
+Often, declarations are bundled together in header files. 
+The most common file extension for header files is `.h`.
+Some projects use `.hpp` or skip the extension completely.
+The definitions are located in a separate `.cpp` file.
+
+
+## Includes
+
+It is possible to include the content of other files with the `include` directive.
+Includes should be stated at the top of the file.
+The snippet below shows how to use the `example()` function that was declared in the `myProject.h` file.
+
+```cpp
+#include "myProject.h"
+example();
+```
+
+If the `myProject.h` is located in the `some_folder` and not in the root directory, you would have to use `#include "some_folder/myProject.h"`.
+
+## Include Guards
+
+It does not matter if the same file is included multiple times within a project.
+Header files should not contain definitions.
+The complete project cannot have the same definition more than once. 
+This is called the "One definition rule".
+It will be enforced by the compiler.
+
+It is easy to avoid multiple unintended inclusions of the same definition with include guards.
+They are formed by a special procedure during the compilation stages.
+The aim is to only include a file only if a certain variable has not been set and then set it once the file is included.
+Often the sequence of this variable is chosen as a variation of the name of the file.
+Another method is the generation of a UUID to reduce the risk of accidentally using the sequence twice.
+The syntax can be seen below with `MY_HEADER_FILE_H` as a variable.
+
+```cpp
+#ifndef MY_HEADER_FILE_H /* any name uniquely mapped to file name */
+#define MY_HEADER_FILE_H
+// file content
+#endif
+```
+A modern form, that will achieve the same result is the #pragma directive `#pragma once`.
+The problem is, that pragmas are not an official part of the C++ language and the implementation differs from compiler to compiler.
+
+## Namespaces
+
+An important method for code organization are namespaces.
+Two functions `int foo()` can be used side-by-side if they are defined in different namespaces.
+They can be called with the scope resolution operator `::`.
+Namespaces can be nested.
+They are kept as they are after includes.
+
+```cpp
+namespace my_ns {
+    int foo() {
+        return 44;
+    }
+    namespace my_inner_ns {
+        int baz() {
+            return 90;
+        }
+    }
+}
+namespace my_other_ns {
+    int foo() {
+        return -2;
+    }
+}
+
+int myresult{my_ns::foo() + my_other_ns::foo() * my_ns::my_inner_ns::baz()};
+```
+
+Deeply nested namespaces might be too verbose.
```

Comment:
nesting namespace will have the problem of relative namespaces
```cpp
namespace bar
{
    int baz() { return 2; }
}

namespace foo {
    namespace bar{
        int baz() { return 1; }
    }
    int which()
    {
        //return bar::baz();
        return ::bar::baz();
    }
}
```

Reply:
also unnamed namespaces help hiding stuff

Reply:
Should we be more specific here? It's already a lot of content. I should talk about the global namespace though.